caikit/core/data_model/base.py

```diff
@@ -70,29 +76,43 @@ def __new__(mcs, name, bases, attrs):
         protobufs, it can be named in the tuple class variable _private_slots and will
         automatically be added to __slots__.
         """
-        # Get all fields in protobufs with same name as class,
-        # except for DataBase, which has no matching protobufs
-        fields = ()
 
-        # Protobufs fields can be divided into these categories, which are used to automatically
-        # determine appropriate behavior in a number of methods
-        fields_enum_map = {}
-        fields_enum_rev = {}
-        _fields_message = ()
-        _fields_message_repeated = ()
-        _fields_enum = ()
-        _fields_enum_repeated = ()
-        _fields_primitive = ()
-        _fields_primitive_repeated = ()
-        _fields_map = ()
+        # Protobufs fields can be divided into these categories, which are used
+        # to automatically determine appropriate behavior in a number of methods
+        attrs["full_name"] = name
+        attrs["fields_enum_map"] = {}
+        attrs["fields_enum_rev"] = {}
+        attrs["_fields_oneofs_map"] = {}
+        attrs["_fields_to_oneof"] = {}
+        attrs["_fields_map"] = ()
+        attrs["_fields_message"] = ()
+        attrs["_fields_message_repeated"] = ()
+        attrs["_fields_enum"] = ()
+        attrs["_fields_enum_repeated"] = ()
+        attrs["_fields_primitive"] = ()
+        attrs["_fields_primitive_repeated"] = ()
+
+        # Look for the set of fields either from a predefined protobuf class or
+        # from a forward declaration from @dataobject
+        fields = ()
         proto_class = None
-        full_name = name
         if name not in ["DataBase", "DataObjectBase"]:
             # Look for a precompiled proto class and if found, parse its
             # descriptor
             proto_class = attrs.get("_proto_class")
             if proto_class is not None:
-                fields = tuple(proto_class.DESCRIPTOR.fields_by_name)
+                all_oneof_fields = [
+                    field.name
+                    for oneof in proto_class.DESCRIPTOR.oneofs
+                    for field in oneof.fields
+                ]
+                fields = tuple(
```

Comment:
This logic here is responsible for ensuring that when using precompiled protos, data is stored using the oneof name and not the individual field name. This is needed so that the right `__slots__` are defined!